reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,147 @@
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val schedulerJob = SupervisorJob()
+
+    private val scope = CoroutineScope(schedulerJob + dispatcher + CoroutineExceptionHandler { _, throwable ->
+        RxJavaPlugins.onError(throwable)
+    })
+
+    override fun scheduleDirect(block: Runnable): Disposable =
+        scheduleDirect(block, 0, TimeUnit.MILLISECONDS)
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (!scope.isActive) return Disposables.disposed()
+        val newBlock = RxJavaPlugins.onSchedule(block)
+        return scope.launch {
+            if (delay > 0) {
+                delay(unit.toMillis(delay))
+            }
+            newBlock.run()
+        }.asDisposable()
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, schedulerJob)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerJob = SupervisorJob(parentJob)
+        private val workerScope = CoroutineScope(workerJob + dispatcher)
+        private val blockChannel = Channel<SchedulerChannelTask>(Channel.UNLIMITED)
+
+        init {
+            workerScope.launch {
+                while (isActive) {
+                    val task = blockChannel.receive()
+                    task.execute()
+                }
+            }
+        }
+
+        override fun isDisposed(): Boolean = !workerScope.isActive
+
+        override fun schedule(block: Runnable): Disposable =
+            schedule(block, 0, TimeUnit.MILLISECONDS)
+
+        override fun schedule(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+            if (!workerScope.isActive) return Disposables.disposed()
+
+            val newBlock = RxJavaPlugins.onSchedule(block)
+
+            fun addTaskToQueue(job: Job) {
+                val task = SchedulerChannelTask(newBlock, job)
+                blockChannel.offer(task)
+            }
+
+            val job = if(delay == 0L) {
+                Job(workerJob).also {
+                    addTaskToQueue(it)
+                }
+            } else {
+                workerScope.launch {
+                    delay(unit.toMillis(delay))
+                    val task = SchedulerChannelTask(newBlock, checkNotNull(coroutineContext[Job]))
+                    blockChannel.offer(task)
+                }
+            }
+
+            return job.asDisposable()
+        }
+
+        override fun dispose() {
+            workerScope.cancel()
+        }
+    }
+}
+
+/**
+ * Represents a task to be queued sequentially on a [Channel] for a [Scheduler.Worker].
+ *
+ * Delayed tasks do not block [Channel] from processing other tasks
+ */
+private class SchedulerChannelTask(
+    private val block: Runnable,
+    job: Job
+) : Disposable {
+    private var isActive = true
```

Comment:
This property is redundant with `job` and actually introduces a whole bunch of issues because of that. Just query/cancel `job` directly.

```kotlin
private class SchedulerChannelTask(
    private val block: Runnable,
    private val job: Job
) : Disposable {
    fun execute() {
        // The task may have been disposed while it was waiting in the queue, so the
        // job's status needs to be explicitly checked before executing it.
        if (job.isActive) {
            block.run()
        }
    }

    override fun isDisposed(): Boolean = !job.isActive

    override fun dispose() {
        job.cancel()
    }
}
```